core/file.go

```diff
@@ -136,6 +138,65 @@ func (file *File) WithTimestamps(ctx context.Context, unix int) (*File, error) {
 	return NewFile(ctx, stamped, path.Base(payload.File), payload.Pipeline, payload.Platform, payload.Services)
 }
 
+func (file *File) Open(
```

Comment:
Is it possible to use the `ReadFile` method on gateway refs to implement an `io.ReadCloser` instead? https://github.com/sipsma/buildkit/blob/3187d2d056de7e3f976ef62cd548499dc3472a7e/frontend/gateway/client/client.go#L101-L101

That method has a `Range` field you can specify, so wouldn't mean we have to just read the whole thing in memory at once I don't think. Advantage would be that we don't have to export the whole image.

And if that's possible, then maybe it's also possible to implement a content.Store using the same idea?

Reply:
Totally! I happen to have [just implemented that somewhere else](https://github.com/vito/bass/blob/76e83c24b7de42e1f3bd83db86ba2c844ae0b62d/pkg/runtimes/util/reffs.go), we could use it here if not just to avoid exporting.

I wonder how it performs vs. a regular export, with all the gRPC overhead. :thinking: Could make sense to explicitly use a buffer that fits just under the message size limit.

> And if that's possible, then maybe it's also possible to implement a content.Store using the same idea?

Ooh, that's a fun idea. Seems theoretically possible considering we'd have arbitrary access. Will look into this.

Reply:
> I wonder how it performs vs. a regular export, with all the gRPC overhead. 🤔 Could make sense to explicitly use a buffer that fits just under the message size limit.

The regular export also has to go over grpc too, so I wouldn't think it would be much different. We do definitely want to be aware of the size limit just so we don't exceed it if nothing else

Reply:
So I looked into this, and we can definitely use a `bkgw.Reference` instead, but it seems too difficult to turn it into a full blown `content.Store` atm.

In a nutshell, we want to be able to access random parts of arbitrary files within single .tar file, the whole point being to avoid having to write an OCI content store to disk, but the bigger point IMO is to support faster feedback loops.

Theoretically this would be easy once you have a mapping from filename to file offsets, but at some point we'd have to transfer the whole file just to find all of these headers. If we're really clever we can do this lazily: keep track of header's we've seen up until offset X, and record more headers as we have to scan further into the file.

In the absolute ideal case Buildkit has already loaded the OCI archive before (e.g. a common image at the bottom of your CI stack). It fetches the image index by walking over a `*tar.Reader` wrapping a `(bkgw.Reference).ReadFile` reader. Ideally the index is right at the start of the file so we don't have to transfer too much data, and then Buildkit doesn't ask for anything else because it sees all the digests are loaded.

In the current case, we will always export the full .tar archive, so there's a fixed cost there. However we only write new content to the store, since the user has a local OCI store cache on disk.

So the ideal case is better, but it really hinges on the index being at the start of the stream, which seems unlikely (unless someone thought about this).

I wanted to implement it anyway, but there isn't a way to get your offset from a `*tar.Reader`, and it buffers, so I can't do it in an inner `io.Reader` either. :( Seems like we would need to implement our own tar header scanner, which seems like a lot more work for possibly not much benefit.

I think it's still worth using a `io.Reader` backed by `bkgw.Reference` though, just to avoid the separate session created by the `Export` call. There just doesn't seem to be a clear path to turning it into a full blown `content.Store`.

Reply:
Switched it over to reading from a `bkfw.Reference`, and added a `core/reffs/` utility package in case we want to use it elsewhere.

Still interested in figuring out the `content.Store` approach someday though, and getting an idea of how often the index is the first file in the stream.